src/Services/Implementations/ReverseGeocodes/BigDataCloudReverseGeocodeService.cs

```diff
@@ -36,14 +44,26 @@ public async Task<IEnumerable<string>> Get(Coordinate coordinate, string? langua
 		return levelNames;
 	}
 
-	public async Task<BigDataCloudResponse?> SerializeFullResponse(Coordinate coordinate, string? language)
+	public async Task<BigDataCloudResponse?> SerializeFullResponse(ReverseGeocodeRequest request)
 	{
 		try
 		{
-			var queryString = $"?latitude={coordinate.Latitude}&longitude={coordinate.Longitude}&key={_apiKeyStore.BigDataCloud}";
-			if (language != null)
-				queryString += $"&localityLanguage={language}";
+			var queryString = $"?latitude={request.Coordinate.Latitude}&longitude={request.Coordinate.Longitude}&key={_apiKeyStore.BigDataCloud}";
+
+			if (request.Language != null)
+			{
+				queryString += $"&localityLanguage={request.Language}";
+			}
+
+			if (_cache.TryGetValue(request, out BigDataCloudResponse? cachedData))
+			{
+				return cachedData;
```

Comment:
This code is very straightforward easily testable by manual tests if the same coordinate bucket is requested from cache. It will be perfect to write an unit test to verify this cache hit logic.

This unit test will also cover your ReverseGeocodeRequest equality and it's hash code implementation.

Creating an new unit test on BigDataCloudReverseGeocodeServiceUnitTests just to cover cache hit & miss scenarios would be perfect.

Reply:
I encountered difficulties testing the code in its current state due to the lack of support for `setup/verification` expressions in the `Dictionary `class, as it is not overridable. Attempting to use `IDictionary` resulted in implementation errors related to dependency injection. Ultimately, I resolved the issue by creating an `ICoordinateCache` interface. This not only addressed [Caching Reverse Geocode Result](https://github.com/photo-cli/photo-cli/issues/5) but also lays the groundwork for a potential evolution into an in-memory database in the future. I'd appreciate hearing your thoughts on this approach. @alpcoker

Reply:
Replied on https://github.com/photo-cli/photo-cli/pull/22#discussion_r1436890861